crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -947,4 +962,27 @@ mod tests {
         assert!(entity.get_by_id(invalid_component_id).is_none());
         assert!(entity.get_mut_by_id(invalid_component_id).is_none());
     }
+
+    // regression test for https://github.com/bevyengine/bevy/pull/7387
+    #[test]
+    fn entity_mut_world_scope_panic() {
+        let mut world = World::new();
+
+        let mut entity = world.spawn_empty();
+        let id = entity.id();
+        let res = std::panic::catch_unwind(AssertUnwindSafe(|| {
+            entity.world_scope(|w| {
+                // Change the entity's `EnityLocation`, which invalidates the original `EntityMut`.
+                // This will get updated at the end of the scope.
+                w.entity_mut(id).insert(TestComponent(0));
+
+                // Ensure that the entity location still gets updated even in case of a panic.
+                panic!("this should get caught by the outer scope")
+            });
+        }));
+        assert!(res.is_err());
+
+        // If the `EntityLocation` has not been updated, this will cause UB.
+        entity.insert(TestComponent(0));
```

Comment:
Would "causing UB" automatically fail the test? Or do we need to be more explicit and check that the location is correct?

Reply:
That's a good point. In this case it does cause the test to fail, but it's could be more rigorous.